### ISSUE TITLE
[Dlang] option -Db_lto=true is ignored

### ISSUE THREAD

**Issue by denizzzka** [mesonbuild/meson#7154]
**Describe the bug**
Option -Db_lto=true (link time optimization) is ignored at least for for ldc2 compiler

**To Reproduce**
Try to run `$ meson setup ---Db_lto=true builddir` and then run `ninja -v`

**system parameters**
* Is this a [cross build](https://mesonbuild.com/Cross-compilation.html) or just a plain native build (for the same computer)?
Both are affected

* what operating system (e.g. MacOS Catalina, Windows 10, CentOS 8.0, Ubuntu 18.04, etc.)
Debian

* what Python version are you using e.g. 3.8.0
Python 3.8.2

* what `meson --version`
~master

* what `ninja --version` if it's a Ninja build
1.10.0

**Comment by dcbaker** [comment]
I've tried enabling this with GDC, I get a linker error (this is with test cases/d/1 simple):
```console
/usr/lib/gcc/x86_64-pc-linux-gnu/9.3.0/include/d/__entrypoint.di:33:5: warning: type of ‘_Dmain’ does not match original declaration [-Wlto-type-mismatch]
   33 | int _Dmain(char[][] args);
      |     ^
../app.d:5:6: note: type mismatch in parameter 1
    5 | void main ()
      |      ^
../app.d:5:6: note: type ‘void’ should match type ‘struct char[][]’
../app.d:5:6: note: ‘D main’ was previously declared here
```

With LDC I can get it to compile, but it can't find LLVMgold.so (which is probably because gentoo provides multiple copies of LLVM so no copy is in the standard LD search path). I know how to fix this, but it requires finding LLVM the dependency, then finding the correct library using the libdir that is provided. Unfortunately that would require a fair amount of messy refactoring, not impossible, just more work than flipping a couple of switches.

**Comment by dcbaker** [comment]
It looks like turning on PGO for ldc has the same issue, because we need to link with clang's runtime (or ldc's if you copied the runtime into the ldc folder.)

**Comment by denizzzka** [comment]
Yes, for Dlang I use a trick - linking with 'c'. (This is another issue, not very important becuse workaround is trivial.)

Current issue is more about passing `--flto` option into Dlang compilers.

**Comment by denizzzka** [comment]
@dcbaker oh, try also add some cross-compile file, in this case clang uses `ld.lld` instead of `gold`
